delegate-app/Main.hs

```diff
@@ -1,49 +1,179 @@
 module Main (main) where
 
 -- Prelude imports
+import Hydra.Cardano.Api.Prelude (lookupEnv)
+import Hydra.Prelude (race_, readMaybe, traverse_)
 import Prelude
 
 -- Haskell imports
+import Control.Concurrent (forkIO, threadDelay)
+import Control.Concurrent.STM (
+  TChan,
+  TQueue,
+  atomically,
+  dupTChan,
+  flushTQueue,
+  newTChan,
+  newTQueue,
+  readTChan,
+  writeTChan,
+  writeTQueue,
+ )
+import Control.Monad (forever, (>=>))
+import Control.Monad.Trans (MonadIO (..), MonadTrans (lift))
+import Control.Tracer (Tracer, contramap, stdoutTracer)
+import Data.Aeson (ToJSON, eitherDecode, encode)
+import Data.Maybe (fromMaybe)
+import Network.WebSockets (
+  acceptRequest,
+  receiveData,
+  runServer,
+  sendTextData,
+  withPingThread,
+ )
+import Prettyprinter (Pretty (pretty))
 
-import Control.Concurrent (threadDelay)
-import Control.Monad.IO.Class (MonadIO (liftIO))
+-- Hydra imports
+import Hydra.Network (IP)
 
 -- HydraAuction imports
 import HydraAuction.Delegate (
   DelegateEvent (Start),
-  DelegateInput (DelegateEvent),
+  DelegateInput (DelegateEvent, FrontendRequest),
   DelegateRunnerT,
   delegateStep,
   execDelegateRunnerT,
  )
+import HydraAuction.Delegate.Interface (DelegateResponse, FrontendRequest)
+import HydraAuction.Delegate.Server (
+  DelegateError (FrontendNoParse),
+  DelegateServerConfig (
+    DelegateServerConfig,
+    dlgt'host,
+    dlgt'ping,
+    dlgt'port,
+    dlgt'tick
+  ),
+  DelegateServerLog (
+    DelegateError,
+    DelegateOutput,
+    FrontendConnected,
+    FrontendInput,
+    Started
+  ),
+  DelegateTracerT,
+  ServerAppT,
+ )
+import HydraAuction.Delegate.Tracing (MonadTracer (trace), TracerT, askTracer, runWithTracer')
+
+{- | per websocket, a thread gets spawned which then
+   subscribes to the broadcast channel and sends to
+   the incoming channel of the delegate runner
+-}
+delegateServerApp ::
+  -- | the amount of seconds between pings
+  Int ->
+  -- | the queue for the inputs of the delegates (this is where the requests are enqueued)
+  TQueue DelegateInput ->
+  -- | the queue for the broadcast (this is where we receives events from the runner)
+  TChan DelegateResponse ->
+  ServerAppT (DelegateTracerT IO)
+delegateServerApp pingSecs reqq broadcast pending = do
+  connection <- liftIO $ acceptRequest pending
+
+  trace FrontendConnected
+  tracer <- askTracer
+
+  liftIO $ do
+    broadcast' <- atomically $ dupTChan broadcast
+
+    let encodeSend :: forall a. ToJSON a => a -> IO ()
+        encodeSend = sendTextData connection . encode
+
+        receive = forever $
```

Comment:
Please move this in `where` blocks or separate functions.

Reply:
These use parameters that are brought in scope in the do block. These are not in scope in a where. They would have to be functions that take two parameters which imho is unnecessarily verbose.

Reply:
Yes. I know. This makes dependencies explicit which is not that bad. Also log do is bad. I tried to explain my position on this here: https://app.slack.com/client/T01UN827FQV/C04C3ASUFEF/thread/C04C3ASUFEF-1680260404.710759

We can leave `let` stuff now and reach consensus on that late.